File: apps/mobile/android/app/src/main/java/com/leadup/nova/NotificationContentGuard.kt
package com.leadup.nova

/**
 * The Kotlin half of NOVA's notification content guard.
 *
 * This mirrors `lib/features/notifications/notification_content_guard.dart`
 * rule-for-rule. The Dart copy is the specification and the one covered by
 * unit tests; this copy exists because it has to run on the path a notification
 * actually takes — inside `onNotificationPosted`, before any text is put on the
 * Flutter event channel, and even when the Flutter engine is not alive at all.
 *
 * The two layers are independent, and a notification is dropped if *either*
 * rejects it, so the effective rule set is the union of the two. That is the
 * intended design: a bug or an omission on one side cannot let an OTP through.
 *
 * Deliberately no logging of the inspected text anywhere in this file. Only
 * package names are ever logged.
 */
internal object NotificationContentGuard {

    /** What kind of never-touch content a notification looks like. */
    enum class Kind {
        ONE_TIME_CODE,
        PASSWORD,
        BANKING,
        AUTH,
    }

    // ── patterns ─────────────────────────────────────────────────────────────

    /** 3-8 digits, optionally grouped by a space or a dash ("123 456"). */
    private val DIGIT_CODE = Regex("""(?<!\d)\d(?:[\s\-]?\d){2,7}(?!\d)""")

    /** `otp` / `o.t.p` / `o t p` / `one time` / `2fa` / `passcode`. */
    private val UNAMBIGUOUS_CODE = Regex(
        """\botp\b|\bo[\s._\-*]?t[\s._\-*]?p\b|\bone[\s\-_]?time\b|\b2fa\b|""" +
            """\btwo[\s\-_]?factor\b|\bmfa\b|\bpasscode\b|\bpass[\s\-]?code\b""",
        RegexOption.IGNORE_CASE,
    )

    /** A code named by what it verifies: "verification code", "login pin". */
    private val QUALIFIED_CODE = Regex(
        """\b(?:verification|confirm(?:ation)?|security|authentication|auth|login|""" +
            """log[\s\-]?in|sign[\s\-]?in|access|activation|invitation|backup|""" +
            """recovery|restore)\s+(?:code|pin|number|token)\b""",
        RegexOption.IGNORE_CASE,
    )

    /** Bare `code` / `pin`, which only counts with a co-signal. */
    private val BARE_CODE = Regex("""\b(?:code|pin)\b""", RegexOption.IGNORE_CASE)

    private val CODE_CO_SIGNAL = Regex(
        """\bdo\s+not\s+(?:share|disclose|reveal)\b|\bnever\s+share\b|""" +
            """\bdon'?t\s+share\b|\bexpires?\s+in\b|\bvalid\s+for\b|""" +
            """\buse\s+(?:this|it|the)\s+code\b|\bcode\s+is\b|\bcode:\s|\bis\s+your\b""",
        RegexOption.IGNORE_CASE,
    )

    private val DO_NOT_SHARE = Regex(
        """\bdo\s+not\s+(?:share|disclose|reveal)\b|\bnever\s+share\b|""" +
            """\bdon'?t\s+share\b""",
        RegexOption.IGNORE_CASE,
    )

    private val PASSWORD = Regex(
        """\bpassword\b|\bpassphrase\b|\bpass\s?phrase\b|\bpasscode\b|""" +
            """\bcredentials?\b|\bmagic\s+link\b|\bpassword\s+reset\b|""" +
            """\breset\s+(?:your\s+)?password\b|\btemporary\s+password\b|""" +
            """\bsign[\s\-]?in\s+link\b|\blogin\s+link\b|""" +
            """\bone[\s\-]?time\s+password\b|\bchange\s+your\s+password\b|""" +
            """\bnew\s+password\b""",
        RegexOption.IGNORE_CASE,
    )

    private val BANKING = Regex(
        """\ba/?c\b|\bacct\b|\bbank\s+account\b|""" +
            """\baccount\s+(?:balance|number|statement|details)\b|\bnet\s?banking\b|""" +
            """\bupi\b|\bneft\b|\bimps\b|\brtgs\b|\bifsc\b|\bmicr\b|\bcvv\b|""" +
            """\bdebit\s+card\b|\bcredit\s+card\b|\bcard\s+ending\b|""" +
            """\bdebited\b|\bcredited\b|\bwithdrawn\b|\bwithdrawal\b|""" +
            """\bdeposited\b|\btransaction\s+(?:of|id|failed|successful|alert)\b|""" +
            """\bemi\s+(?:due|of|amount)\b|\bbalance\s+is\b|\bbank\s+alert\b|""" +
            """\bstatement\s+is\s+(?:ready|generated)\b|\bspent\s+(?:on|of|at)\b|""" +
            """\bpaid\s+to\b|\breceived\s+from\b|\bupi\s+(?:id|ref|transaction)\b""",
        RegexOption.IGNORE_CASE,
    )

    /** An escaped normal string: `$` and `\d` cannot go in a raw Kotlin string. */
    private val CURRENCY = Regex(
        "(?:₹|rs\\.?\\s?|inr\\s?|usd\\s?|\\$|€|eur\\s?|£|gbp\\s?|aed\\s?|sgd\\s?)\\d" +
            "|\\d+(?:[.,]\\d+)?\\s?(?:inr|usd|eur|gbp|aed|sgd|rupees|dollars|euros)\\b",
        RegexOption.IGNORE_CASE,
    )

    private val MONEY_CONTEXT = Regex(
        """\bpaid\b|\bpay(?:ment)?\b|\bspent\b|\bcharged\b|\bdebit(?:ed)?\b|""" +
            """\bcredit(?:ed)?\b|\btransferred\b|\brefund\b|\bbalance\b|\binvoice\b|""" +
            """\bemi\b|\btransaction\b|\bwithdraw(?:al|n)?\b|\bdeposit(?:ed)?\b""",
        RegexOption.IGNORE_CASE,
    )

    private val AUTH = Regex(
        """\bverify\s+your\b|\bconfirm\s+your\s+(?:email|account|identity)\b|""" +
            """\bverification\s+(?:email|link|request)\b|\bsign[\s\-]?in\s+attempt\b|""" +
            """\bsign[\s\-]?in\s+request\b|\bnew\s+(?:device|login|sign[\s\-]?in)\b|""" +
            """\b(?:approve|deny|confirm)\s+(?:the\s+|this\s+)?""" +
            """(?:sign[\s\-]?in|login|log[\s\-]?in|authentication|access)\b|""" +
            """\bsecurity\s+alert\b|\bunusual\s+(?:sign[\s\-]?in|activity|login)\b|""" +
            """\bsuspicious\s+(?:login|activity|sign[\s\-]?in)\b|""" +
            """\bauthentication\s+(?:code|request|prompt)\b|\baccess\s+code\b|""" +
            """\bone[\s\-]?time\s+(?:code|pin|password)\b""",
        RegexOption.IGNORE_CASE,
    )

    /**
     * Package-name fragments that mark an app as financial or
     * credential-bearing.
     *
     * Mirrors the Dart hint table minus the entries covered by the explicit
     * catalogue there. Deliberately conservative: a false positive makes an app
     * un-selectable, so broad tokens such as `pay` (Payroll, Paycom) are left
     * out.
     */
    private val TREASURY_PACKAGE_HINTS = mapOf(
        "bank" to Kind.BANKING,
        "wallet" to Kind.BANKING,
        "paypal" to Kind.BANKING,
        "paytm" to Kind.BANKING,
        "phonepe" to Kind.BANKING,
        "paisa" to Kind.BANKING,
        "upi" to Kind.BANKING,
        "authenticator" to Kind.ONE_TIME_CODE,
        "authy" to Kind.ONE_TIME_CODE,
        "freeotp" to Kind.ONE_TIME_CODE,
        "aegis" to Kind.ONE_TIME_CODE,
        "password" to Kind.PASSWORD,
        "passcode" to Kind.PASSWORD,
        "bitwarden" to Kind.PASSWORD,
        "lastpass" to Kind.PASSWORD,
        "dashlane" to Kind.PASSWORD,
        "onepassword" to Kind.PASSWORD,
    )

    /** Whole words that mark an installed app's label as financial or secret. */
    private val TREASURY_LABEL_HINTS = mapOf(
        "bank" to Kind.BANKING,
        "banks" to Kind.BANKING,
        "banking" to Kind.BANKING,
        "netbanking" to Kind.BANKING,
        "passbook" to Kind.BANKING,
        "wallet" to Kind.BANKING,
        "authenticator" to Kind.ONE_TIME_CODE,
        "otp" to Kind.ONE_TIME_CODE,
        "password" to Kind.PASSWORD,
        "passcode" to Kind.PASSWORD,
    )

    private val TREASURY_LABEL_HINT = Regex(
        "\\b(?:" +
            TREASURY_LABEL_HINTS.keys
                .sortedByDescending { it.length }
                .joinToString("|") +
            ")\\b",
        RegexOption.IGNORE_CASE,
    )

    // ── rules ────────────────────────────────────────────────────────────────

    /**
     * The §9.5 non-negotiable: does this text look like an OTP, a password, a
     * bank alert or an authentication message?
     *
     * Returns null when the text is not recognisably sensitive. The caller
     * treats any non-null answer as "drop, store nothing, say nothing".
     */
    fun detect(title: String, body: String): Kind? {
        val raw = "$title\n$body".lowercase()
        if (raw.isBlank()) return null

        // Leet-normalised copy used only for keyword matching; the digit-based
        // patterns run against the untouched text so this cannot invent a code.
        val words = deLeet(raw)

        if (BANKING.containsMatchIn(raw) || BANKING.containsMatchIn(words)) {
            return Kind.BANKING
        }
        if (CURRENCY.containsMatchIn(raw) && MONEY_CONTEXT.containsMatchIn(words)) {
            return Kind.BANKING
        }
        if (PASSWORD.containsMatchIn(words)) return Kind.PASSWORD
        if (oneTimeCode(raw, words)) return Kind.ONE_TIME_CODE
        if (AUTH.containsMatchIn(words)) return Kind.AUTH
        return null
    }

    /**
     * The permanent block, mirroring `blockedCategoryFor` in
     * `notification_app_catalogue.dart`.
     *
     * [label] is the installed app's label from `PackageManager`, never a string
     * the notification supplied, so it cannot be spoofed by the notification
     * payload.
     */
    fun blockedCategoryFor(packageName: String, label: String?): Kind? {
        val lower = packageName.lowercase()
        for ((hint, kind) in TREASURY_PACKAGE_HINTS) {
            if (lower.contains(hint)) return kind
        }
        val text = label?.trim().orEmpty()
        if (text.isEmpty()) return null
        val match = TREASURY_LABEL_HINT.find(text) ?: return null
        return TREASURY_LABEL_HINTS[match.value.lowercase()]
    }

    private fun oneTimeCode(raw: String, words: String): Boolean {
        if (DIGIT_CODE.containsMatchIn(raw)) {
            if (UNAMBIGUOUS_CODE.containsMatchIn(words)) return true
            if (QUALIFIED_CODE.containsMatchIn(words)) return true
            if (BARE_CODE.containsMatchIn(raw) && CODE_CO_SIGNAL.containsMatchIn(raw)) {
                return true
            }
        }
        return DO_NOT_SHARE.containsMatchIn(raw) &&
            (BARE_CODE.containsMatchIn(raw) || UNAMBIGUOUS_CODE.containsMatchIn(words))
    }

    /** `O.T.P` → `otp`, `0tp` → `otp`, `pa55word` → `password`. */
    private fun deLeet(input: String): String {
        val substitutions = mapOf(
            '0' to 'o',
            '1' to 'l',
            '3' to 'e',
            '4' to 'a',
            '5' to 's',
            '7' to 't',
            '@' to 'a',
            '$' to 's',
        )
        val builder = StringBuilder(input.length)
        for (char in input) {
            builder.append(substitutions[char] ?: char)
        }
        return builder.toString()
    }
}

File: apps/mobile/lib/features/notifications/notification_content_guard.dart
/// What kind of never-touch content a notification looks like.
enum SensitiveContentKind {
  /// A one-time passcode, verification code or authenticator prompt.
  oneTimeCode,

  /// A password, passphrase, magic link or credential.
  passwordOrCredential,

  /// A bank, card, UPI or transaction alert.
  bankingAlert,

  /// A sign-in / account-verification message with no code in it.
  authenticationMessage,
}

/// The §9.5 non-negotiable: recognise notification content that must never be
/// stored, relayed or spoken.
///
/// This is deliberately one named function over one set of named patterns
/// rather than conditionals scattered through the filter, the controller and
/// the UI. `NotificationFilter` calls it, the read-aloud path calls it again
/// immediately before speaking, and `NotificationContentGuard.kt` mirrors it on
/// the Kotlin side so the check also runs when the Flutter engine is not alive.
///
/// ## Why it errs toward dropping
///
/// Any app can post a notification with any text, so the input is hostile. The
/// cost of a false positive is a work summary the user does not see; the cost
/// of a false negative is an OTP in a log, a database or a speaker. Every
/// ambiguity therefore resolves to "sensitive".
///
/// ## What it does and does not claim
///
/// It is a heuristic, not a classifier. It will not catch a code with no
/// keyword anywhere near it ("482913"). That case is covered by the app-level
/// blocklist for OTP/authenticator packages, and it is exactly why §5.21 makes
/// "Do not store raw notification text" a rule rather than relying on
/// detection alone.
abstract final class NotificationContentGuard {
  /// Lower-cased text with separator-obfuscated keywords normalised, so `O.T.P`
  /// and `0TP` are recognised as `otp`.
  static SensitiveContentKind? detect({required String title, required String body}) {
    final raw = '$title\n$body'.toLowerCase();
    if (raw.trim().isEmpty) return null;

    // Leet-normalised copy used only for *keyword* matching. The digit-based
    // code patterns below run against the untouched text, so normalisation
    // cannot turn a code into a keyword.
    final words = _deLeet(raw);

    if (_bankingKeyword.hasMatch(raw) || _bankingKeyword.hasMatch(words)) {
      return SensitiveContentKind.bankingAlert;
    }
    if (_currencyAmount.hasMatch(raw) && _moneyContext.hasMatch(words)) {
      return SensitiveContentKind.bankingAlert;
    }
    if (_passwordKeyword.hasMatch(words)) {
      return SensitiveContentKind.passwordOrCredential;
    }
    if (_oneTimeCode(raw, words)) {
      return SensitiveContentKind.oneTimeCode;
    }
    if (_authenticationKeyword.hasMatch(words)) {
      return SensitiveContentKind.authenticationMessage;
    }
    return null;
  }

  /// Convenience wrapper for the read-aloud path, where only the already
  /// sanitised body is available.
  static SensitiveContentKind? detectText(String text) =>
      detect(title: '', body: text);

  // ── patterns ─────────────────────────────────────────────────────────────

  /// A run of 3-8 digits, optionally grouped by a single space or dash so
  /// "123 456" and "123-456" are seen as one code.
  static final RegExp _digitCode = RegExp(r'(?<!\d)\d(?:[\s\-]?\d){2,7}(?!\d)');

  /// `otp` / `o.t.p` / `o t p` / `one time` / `one-time` / `2fa` / `two factor`
  /// / `mfa` / `passcode` — an authentication code named unambiguously.
  static final RegExp _unambiguousCodeKeyword = RegExp(
    r'\botp\b|\bo[\s._\-*]?t[\s._\-*]?p\b|\bone[\s\-_]?time\b|\b2fa\b|'
    r'\btwo[\s\-_]?factor\b|\bmfa\b|\bpasscode\b|\bpass[\s\-]?code\b',
  );

  /// A code named by the thing it verifies: "verification code", "security
  /// pin", "login code", "access code", "activation code".
  static final RegExp _qualifiedCodeKeyword = RegExp(
    r'\b(?:verification|confirm(?:ation)?|security|authentication|auth|login|'
    r'log[\s\-]?in|sign[\s\-]?in|access|activation|invitation|backup|recovery|'
    r'restore)\s+(?:code|pin|number|token)\b',
  );

  /// Bare `code` / `pin` / `number`, which only counts with a co-signal.
  static final RegExp _bareCodeKeyword = RegExp(r'\b(?:code|pin)\b');

  /// The co-signals that make a bare `code` an authentication code rather than
  /// an order reference. `don't share` is written with `\u0027` so the pattern
  /// stays a single-quoted Dart string.
  static final RegExp _codeCoSignal = RegExp(
    r'\bdo\s+not\s+(?:share|disclose|reveal)\b|\bnever\s+share\b|'
    r'\bdon\u0027?t\s+share\b|\bexpires?\s+in\b|\bvalid\s+for\b|'
    r'\buse\s+(?:this|it|the)\s+code\b|\bcode\s+is\b|\bcode:\s|\bis\s+your\b',
  );

  static final RegExp _doNotShare = RegExp(
    r'\bdo\s+not\s+(?:share|disclose|reveal)\b|\bnever\s+share\b|'
    r'\bdon\u0027?t\s+share\b',
  );

  /// Passwords and credentials, in the shapes they actually arrive in.
  static final RegExp _passwordKeyword = RegExp(
    r'\bpassword\b|\bpassphrase\b|\bpass\s?phrase\b|\bpasscode\b|'
    r'\bcredentials?\b|\bmagic\s+link\b|\bpassword\s+reset\b|'
    r'\breset\s+(?:your\s+)?password\b|\btemporary\s+password\b|'
    r'\bsign[\s\-]?in\s+link\b|\blogin\s+link\b|\bone[\s\-]?time\s+password\b|'
    r'\bchange\s+your\s+password\b|\bnew\s+password\b',
  );

  /// Bank, card and transaction language. Every entry is unambiguous enough to
  /// stand alone; generic finance words that also describe ordinary work
  /// ("invoice", "payment", "amount") are handled by the currency rule below.
  ///
  /// `credit` and `debit` appear on their own only in their inflected forms:
  /// the bare nouns are common in ordinary work prose ("credit the team") and
  /// would drop legitimate messages.
  static final RegExp _bankingKeyword = RegExp(
    r'\ba/?c\b|\bacct\b|\bbank\s+account\b|'
    r'\baccount\s+(?:balance|number|statement|details)\b|\bnet\s?banking\b|'
    r'\bupi\b|\bneft\b|\bimps\b|\brtgs\b|\bifsc\b|\bmicr\b|\bcvv\b|'
    r'\bdebit\s+card\b|\bcredit\s+card\b|\bcard\s+ending\b|'
    r'\bdebited\b|\bcredited\b|\bwithdrawn\b|\bwithdrawal\b|'
    r'\bdeposited\b|\btransaction\s+(?:of|id|failed|successful|alert)\b|'
    r'\bemi\s+(?:due|of|amount)\b|\bbalance\s+is\b|\bbank\s+alert\b|'
    r'\bstatement\s+is\s+(?:ready|generated)\b|\bspent\s+(?:on|of|at)\b|'
    r'\bpaid\s+to\b|\breceived\s+from\b|\bupi\s+(?:id|ref|transaction)\b',
  );

  static final RegExp _currencyAmount = RegExp(
    r'(?:₹|rs\.?\s?|inr\s?|usd\s?|\$|€|eur\s?|£|gbp\s?|aed\s?|sgd\s?)\d'
    r'|\d+(?:[.,]\d+)?\s?(?:inr|usd|eur|gbp|aed|sgd|rupees|dollars|euros)\b',
  );

  static final RegExp _moneyContext = RegExp(
    r'\bpaid\b|\bpay(?:ment)?\b|\bspent\b|\bcharged\b|\bdebit(?:ed)?\b|'
    r'\bcredit(?:ed)?\b|\btransferred\b|\brefund\b|\bbalance\b|\binvoice\b|'
    r'\bemi\b|\btransaction\b|\bwithdraw(?:al|n)?\b|\bdeposit(?:ed)?\b',
  );

  /// Sign-in and account-security messages that carry no code.
  ///
  /// `approve`/`deny` only count next to an authentication noun, so an ordinary
  /// "approve the design doc" is not swallowed by this rule.
  static final RegExp _authenticationKeyword = RegExp(
    r'\bverify\s+your\b|\bconfirm\s+your\s+(?:email|account|identity)\b|'
    r'\bverification\s+(?:email|link|request)\b|\bsign[\s\-]?in\s+attempt\b|'
    r'\bsign[\s\-]?in\s+request\b|\bnew\s+(?:device|login|sign[\s\-]?in)\b|'
    r'\b(?:approve|deny|confirm)\s+(?:the\s+|this\s+)?'
    r'(?:sign[\s\-]?in|login|log[\s\-]?in|authentication|access)\b|'
    r'\bsecurity\s+alert\b|\bunusual\s+(?:sign[\s\-]?in|activity|login)\b|'
    r'\bsuspicious\s+(?:login|activity|sign[\s\-]?in)\b|'
    r'\bauthentication\s+(?:code|request|prompt)\b|\baccess\s+code\b|'
    r'\bone[\s\-]?time\s+(?:code|pin|password)\b',
  );

  // ── rules ────────────────────────────────────────────────────────────────

  /// Whether the text carries a one-time code.
  ///
  /// Three independent signals, any of which is enough:
  ///  1. a code named unambiguously (`otp`, `one-time`, `2fa`, `passcode`) plus
  ///     any digit run;
  ///  2. a code named by what it verifies (`verification code`, `login pin`)
  ///     plus any digit run;
  ///  3. a bare `code`/`pin` plus a co-signal ("do not share", "expires in",
  ///     "code is") plus a digit run;
  ///  4. a "do not share" instruction plus a code word, even with no digits —
  ///     the sentence itself is an OTP warning.
  static bool _oneTimeCode(String raw, String words) {
    final hasDigits = _digitCode.hasMatch(raw);
    if (hasDigits) {
      if (_unambiguousCodeKeyword.hasMatch(words)) return true;
      if (_qualifiedCodeKeyword.hasMatch(words)) return true;
      if (_bareCodeKeyword.hasMatch(raw) && _codeCoSignal.hasMatch(raw)) {
        return true;
      }
    }
    return _doNotShare.hasMatch(raw) &&
        (_bareCodeKeyword.hasMatch(raw) ||
            _unambiguousCodeKeyword.hasMatch(words));
  }

  /// `O.T.P` → `otp`, `0tp` → `otp`, `c0de` → `code`, `pa55word` → `password`.
  ///
  /// Only applied to the *keyword* copy. Deliberate obfuscation of an OTP
  /// keyword is a real pattern in bulk SMS and in apps that try to slip past
  /// on-device filters, and normalising is cheap. Because the digit-based
  /// patterns above run against the untouched text, this cannot invent a code.
  static String _deLeet(String input) {
    const substitutions = <String, String>{
      '0': 'o',
      '1': 'l',
      '3': 'e',
      '4': 'a',
      '5': 's',
      '7': 't',
      '@': 'a',
      r'$': 's',
    };
    final buffer = StringBuffer();
    for (final rune in input.runes) {
      final char = String.fromCharCode(rune);
      buffer.write(substitutions[char] ?? char);
    }
    return buffer.toString();
  }
}
